extensions/binding/org.eclipse.smarthome.binding.mqttbroker/src/main/java/org/eclipse/smarthome/binding/mqttbroker/internal/MqttWillAndTestamentBuilder.java

```diff
@@ -0,0 +1,65 @@
+/**
+ * Copyright (c) 2010-2017 by the respective copyright holders.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.smarthome.binding.mqttbroker.internal;
+
+import org.apache.commons.lang.StringUtils;
+import org.eclipse.smarthome.io.transport.mqtt.MqttWillAndTestament;
+
+/**
+ * Builds a {@link MqttWillAndTestament}.
+ *
+ * @author David Graeff - Initial contribution
+ */
+public class MqttWillAndTestamentBuilder {
+    private String topic;
+    private byte[] message;
+    private int qos;
+    private boolean retain;
+
+    public static MqttWillAndTestamentBuilder create(String topic) {
+        if (StringUtils.isBlank(topic)) {
+            throw new IllegalArgumentException("Topic must not be null");
+        }
+        MqttWillAndTestamentBuilder b = new MqttWillAndTestamentBuilder();
+        b.topic = topic;
+        return b;
+    }
+
+    public MqttWillAndTestamentBuilder qos(Integer qos) {
+        if (qos != null) {
+            if (qos < 0 || qos > 2) {
+                throw new IllegalArgumentException("QOS value invalid");
+            }
+            this.qos = qos;
+        }
+        return this;
+    }
+
+    public MqttWillAndTestamentBuilder retain(Boolean retain) {
+        if (retain != null) {
+            this.retain = retain;
+        }
+        return this;
+    }
+
+    public MqttWillAndTestamentBuilder payload(String payload) {
+        if (payload != null) {
+            this.message = payload.getBytes();
+        }
+        return this;
+    }
+
+    public MqttWillAndTestament build() {
+        try {
+            return new MqttWillAndTestament(topic, message, qos, retain);
+        } catch (IllegalArgumentException ignored) {
+            return null;
```

Comment:
Hm, is this really the best way to handle it?

Reply:
Yes, if it fails, we return null. At least at the moment that's the contract. The builder could throw as well, but the consumer can use the MqttWillAndTestament constructor if the exact error knowledge and full control is the requirement.